### PR TITLE
Add Realm backup \ world download endpoints

### PR DESCRIPTION
**Pull request by LucienHH**

### CODE REVIEW

File: test/common/responses.json [hunk 1]
```diff
@@ -68,5 +68,33 @@
...
+        "backups": [
+            {
+                "backupId": "1970-01-01T00:00:00.000Z",
+                "lastModifiedDate": 1652129181690,
+                "size": 10000,
+                "metadata": {
+                    "game_difficulty": "0",
+                    "name":"Test Realm",
+                    "game_server_version": "1.19.0",
+                    "enabled_packs": "{\"resourcePacks\":[],\"behaviorPacks\":[\"4b45f411-9ca2-4d12-a2a8-4d2de8746538\"]}",
```
Comment (extremeheat) [inline test/common/responses.json:82]: we should abstract this, at least by JSON.parsing() it

File: src/structures/Backup.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  #api
+  #realm
+  constructor (api, realm, data) {
+    this.#api = api
+    this.#realm = realm
+    Object.assign(this, {
+      id: data.backupId,
+      lastModifiedDate: data.lastModifiedDate,
+      size: data.size,
+      metadata: {
```
Comment (extremeheat) [inline src/structures/Backup.js:11]: is there a reason to encapsulate in metadata subfield ?

File: src/index.js [hunk 3]
```diff
@@ -22,6 +25,15 @@ class RealmAPI {
...
     return data.servers.map(realm => new Realm(this, realm))
   }
+
+  async getRealmBackups (realmId, slotId) {
+    const data = await this.rest.get(`/worlds/${realmId}/backups`)
+    return data.backups.map(e => new Backup(this, { realmId, slotId }, e))
+  }
+
+  async restoreRealmFromBackup (realmId, backupId) {
+    return await this.rest.put(`/worlds/${realmId}/backups?backupId=${backupId}&clientSupportsRetries`, { retryCount: 5 })
```
Comment (extremeheat) [inline src/index.js:35]: if backupId can contain arbitrary characters, then this call is unsafe. You need to use encodeURIComponent.

File: src/structures/Realm.js [hunk 4]
```diff
@@ -44,4 +44,12 @@ module.exports = class Realm {
   async close () {
     return this.#api.changeRealmState(this.id, 'close')
   }
+
+  async getWorldDownload () {
+    return this.#api.getRealmWorldDownload(this.id, this.activeSlot, 'latest')
```
Comment (extremeheat) [inline src/structures/Realm.js:49]: what is activeSlot ? I don't see it in the docs

File: src/structures/Download.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      this.resourcePackHash = data.resourcePackHash
+      this.fileExtension = '.tar.gz'
+    }
+  }
+
+  async toDirectory (directory) {
+    return this.#downloadWorld().then(buffer => fs.writeFile(`${directory}/world${this.fileExtension}`, buffer))
+  }
+
+  async toBuffer () {
```
Comment (extremeheat) [inline src/structures/Download.js:24]: for clarity, rename the methods to `getBuffer()` and `writeToDirectory()`. Also, use fs promises and only return and resolve once the download and write to disk is complete.

File: src/java/api.js [hunk 6]
```diff
@@ -21,4 +23,9 @@ module.exports = class JavaRealmAPI extends RealmAPI {
   async changeRealmState (realmId, state) {
     return await this.rest.put(`/worlds/${realmId}/${state}`)
   }
+
+  async getRealmWorldDownload (realmId, slotId) {
```
Comment (extremeheat) [inline src/java/api.js:27]: I'm confused as to why `Download` can't be inlined into `Backup`? Is this getRealmWorldDownload() not using backups and calling a seperate API?

**Review by extremeheat — CHANGES_REQUESTED** [review]
Left some comments

File: src/java/api.js (hunk 6 above)
Comment (LucienHH) [reply src/java/api.js:27]: Yeah that's correct, on Java you cannot download individual backups only the world at the time the request is called.

File: src/structures/Backup.js (hunk 2 above)
Comment (LucienHH) [reply src/structures/Backup.js:11]: It's as it was returned from the API and think it fits as doesn't relate to the backup but the metadata for the world.

File: src/structures/Realm.js (hunk 4 above)
Comment (LucienHH) [reply src/structures/Realm.js:49]: Realms have 3 slots for worlds, activeSlot is the currently selected slot and is a property of `Realm`

File: src/structures/Download.js (hunk 5 above)
Comment (LucienHH) [reply src/structures/Download.js:24]: Will do, it already uses fs promsises and does only return once the downloads done.

File: docs/API.md [hunk 7]
```diff
@@ -363,3 +430,153 @@ await realm.close()
...
+    getDownload(): Promise<Download>
+    restore(): Promise<void>
+    id: string
+    lastModifiedDate: number
+    size: number
+    metadata: {
+        gameDifficulty: string
+        name: string
+        gameServerVersion: string
+        enabledPacks: string
```
Comment (extremeheat) [inline docs/API.md:487]: Docs for this field are outdated. You can either inline the fields into the Backup object or update the type here to be an object containing { resourcePack, behaviorPack }

Comment (LucienHH) [reply docs/API.md]: Fixed

**Comment by rom1504** [comment]
is this ready

**Comment by LucienHH** [comment]
I believe so, @extremeheat any further suggestions?

**Comment by extremeheat** [comment]
Not tested personally, but I assume it works so LGTM

**Review by extremeheat — APPROVED** [review]

**Comment by LucienHH** [comment]
@rom1504 I think this is all good to go?

**Comment by extremeheat** [comment]
thanks!

**Comment by extremeheat** [comment]
You can open a release PR

**Pull request merged**